src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -133,7 +134,71 @@ internal void SendPendingBatches()
             // Dispatch any buffered renders we accumulated during a disconnect.
             // Note that while the rendering is async, we cannot await it here. The Task returned by ProcessBufferedRenderBatches relies on
             // OnRenderCompleted to be invoked to complete, and SignalR does not allow concurrent hub method invocations.
-            var _ = Renderer.Dispatcher.InvokeAsync(() => Renderer.ProcessBufferedRenderBatches());
+            _ = Renderer.Dispatcher.InvokeAsync(() => Renderer.ProcessBufferedRenderBatches());
+        }
+
+        internal async Task EndInvokeJSFromDotNet(long asyncCall, bool succeded, string arguments)
+        {
+            try
+            {
+                AssertInitialized();
+
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+                    if (!succeded)
+                    {
+                        // We can log the arguments here because it is simply the JS error with the call stack.
+                        Log.EndInvokeJSFailed(_logger, asyncCall, arguments);
+                    }
+                    else
+                    {
+                        Log.EndInvokeJSSucceeded(_logger, asyncCall);
+                    }
+
+                    DotNetDispatcher.EndInvoke(arguments);
+                });
+            }
+            catch (Exception ex)
+            {
+                Log.EndInvokeDispatchException(_logger, ex);
+            }
+        }
+
+        internal async Task DispatchEvent(string eventDescriptorJson, string eventArgs)
+        {
+            try
+            {
+                AssertInitialized();
+
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+
+                    var eventDescriptor = ParseEventDescriptor(eventDescriptorJson);
```

Comment:
Could we move the parsing out of the `InvokeAsync` call, so that:

1. We're freeing up more of the sync context's time to do things that have to be done inside the sync context (super minor, I know)
2. More importantly, we can have a separate `catch` so that we can differentiate parsing failures from actual event-handling logic failures. Then the log message for exceptions from `RendererRegistryEventDispatcher.DispatchEvent` can specify the event handler ID. This will help with diagnosing cases like "that event handler ID was already disposed".

Reply:
Thought we would already log these things as part of the work you did on the renderer. I added the EventHandlerId, but you should consider adding those logs to the renderer itself.